cmd/gateway/gcs/gateway-gcs.go

```diff
@@ -957,14 +962,165 @@ func (l *gcsGateway) NewMultipartUpload(ctx context.Context, bucket string, key
+		}
+	}
+
+	if commonPrefix != "" {
+		return nil, commonPrefix
+	}
+
+	return &minio.MultipartInfo{
+		Object:    mpMeta.Object,
+		UploadID:  components[3],
+		Initiated: attrs.Created,
+	}, ""
+}
+
+// To sort uploads by key followed by upload initiated time.
+type byGCSUploads []minio.MultipartInfo
+
+func (b byGCSUploads) Len() int {
+	return len(b)
+}
+
+func (b byGCSUploads) Less(i, j int) bool {
+	if b[i].Object == b[j].Object {
+		return b[i].Initiated.Sub(b[j].Initiated) < 0
+	}
+	return b[i].Object < b[j].Object
+}
+
+func (b byGCSUploads) Swap(i, j int) {
+	b[i], b[j] = b[j], b[i]
+}
+
 // ListMultipartUploads - lists all multipart uploads.
 func (l *gcsGateway) ListMultipartUploads(ctx context.Context, bucket string, prefix string, keyMarker string, uploadIDMarker string, delimiter string, maxUploads int) (minio.ListMultipartsInfo, error) {
+	// List objects under <bucket>/gcsMinioMultipartPathV1
+	it := l.client.Bucket(bucket).Objects(ctx, &storage.Query{
+		Prefix: gcsMinioMultipartPathV1,
+	})
+
+	var (
+		uploads  []minio.MultipartInfo
+		prefixes []string
+		count    int
+	)
+
+	isTruncated := true
+	for count < maxUploads {
+		attrs, err := it.Next()
+		if err == iterator.Done {
+			isTruncated = false
+			break
+		}
+
+		if err != nil {
+			logger.LogIf(ctx, err)
+			return minio.ListMultipartsInfo{
+				KeyMarker:      keyMarker,
+				UploadIDMarker: uploadIDMarker,
+				MaxUploads:     maxUploads,
+				Prefix:         prefix,
+				Delimiter:      delimiter,
+			}, gcsToObjectError(err)
+		}
+
+		mpartInfo, commonPrefix := l.gcsGetUploadID(ctx, attrs, bucket, prefix)
+		if mpartInfo == nil && commonPrefix == "" {
+			continue
+		}
+
+		if mpartInfo != nil {
+			if keyMarker != "" && uploadIDMarker == "" &&
+				mpartInfo.Object <= keyMarker {
+				continue
+			}
+
+			if keyMarker != "" && uploadIDMarker != "" &&
+				mpartInfo.Object < keyMarker &&
+				mpartInfo.UploadID <= uploadIDMarker {
+				continue
+			}
+
+			uploads = append(uploads, *mpartInfo)
+			count++
+		}
+
+		if commonPrefix != "" {
+			prefixes = append(prefixes, commonPrefix)
+		}
+	}
+
+	// Sort uploads by key followed by initiated time
+	sort.Sort(byGCSUploads(uploads))
+
+	nextKeyMarker := ""
+	nextUploadIDMarker := ""
+
+	if isTruncated {
+		lastUploadIndex := maxUploads - 1
+		nextKeyMarker = uploads[lastUploadIndex].Object
+		nextUploadIDMarker = uploads[lastUploadIndex].UploadID
+	}
+
 	return minio.ListMultipartsInfo{
-		KeyMarker:      keyMarker,
-		UploadIDMarker: uploadIDMarker,
-		MaxUploads:     maxUploads,
-		Prefix:         prefix,
-		Delimiter:      delimiter,
+		KeyMarker:          keyMarker,
+		UploadIDMarker:     uploadIDMarker,
+		MaxUploads:         maxUploads,
+		Prefix:             prefix,
+		Delimiter:          delimiter,
+		Uploads:            uploads,
+		CommonPrefixes:     prefixes,
+		NextKeyMarker:      nextKeyMarker,
+		NextUploadIDMarker: nextUploadIDMarker,
+		IsTruncated:        isTruncated,
```

Comment:
GCS backend format was not designed to implement ListMultipartUploads, there is no need for supporting CommonPrefixes, marker etc. We can do just enough to support docker registry. i.e the following implementation will be good enough (and we need not complicate our code):

```
// ListMultipartUploads - lists all multipart uploads.
func (l *gcsGateway) ListMultipartUploads(ctx context.Context, bucket string, prefix string, keyMarker string, uploadIDMarker string, delimiter string, maxUploads int) (minio.ListMultipartsInfo, error) {
        // List objects under <bucket>/gcsMinioMultipartPathV1
        it := l.client.Bucket(bucket).Objects(ctx, &storage.Query{
                Prefix: gcsMinioMultipartPathV1,
        })

        var uploads []minio.MultipartInfo

        for {
                attrs, err := it.Next()
                if err == iterator.Done {
                        break
                }

                if err != nil {
                        logger.LogIf(ctx, err)
                        return minio.ListMultipartsInfo{
                                KeyMarker:      keyMarker,
                                UploadIDMarker: uploadIDMarker,
                                MaxUploads:     maxUploads,
                                Prefix:         prefix,
                                Delimiter:      delimiter,
                        }, gcsToObjectError(err)
                }

                // Extract multipart upload information from gcs.json
                obj := l.client.Bucket(bucket).Object(attrs.Name)
                objReader, rErr := obj.NewReader(ctx)
                if rErr != nil {
                        logger.LogIf(ctx, rErr)
                        return nil, ""
                }
                defer objReader.Close()

                var mpMeta gcsMultipartMetaV1
                dec := json.NewDecoder(objReader)
                decErr := dec.Decode(&mpMeta)
                if decErr != nil {
                        logger.LogIf(ctx, decErr)
                        return nil, ""
                }
                if prefix == mpMeta.Object {
                        // Extract uploadId
                        // E.g minio.sys.tmp/multipart/v1/d063ad89-fdc4-4ea3-a99e-22dba98151f5/gcs.json
                        components := strings.SplitN(attrs.Name, "/", 5)
                        if len(components) != 5 {
                                logger.LogIf(ctx, errors.New("Invalid multipart upload format"))
                                return nil, ""
                        }
                        upload := minio.MultipartInfo{mpMeta.Object, components[3], attrs.Created}
                        uploads = []minio.MultipartInfo{upload}
                        break
                }
        }

        return minio.ListMultipartsInfo{
                KeyMarker:          keyMarker,
                UploadIDMarker:     uploadIDMarker,
                MaxUploads:         maxUploads,
                Prefix:             prefix,
                Delimiter:          delimiter,
                Uploads:            uploads,
                CommonPrefixes:     prefixes,
                NextKeyMarker:      "",
                NextUploadIDMarker: "",
                IsTruncated:        false,
        }, nil
}
```

Reply:
Done